calblueprint/dccentralkitchen
# Auth redesign and refactor

[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

# What's new in this PR
This PR completely revamps the auth flow to rely on phone number verification which eliminates the need for passwords and significantly simplifies the entire auth flow into a single path.

**New Process Overview**
`OnboardingScreen` is now the only opening screen.
 - Selecting **Get Started** navigates to `PhoneNumberScreen` where the user enters their phone number. The user enters and verifies their phone number (`VerificationScreen`) regardless of if they have an account.
   - If there exists an account with their phone number, they're taken directly into the app after verifying the phone number.
   - If there is not an account with their phone number, they're taken to `CompleteSignUpScreen` where they enter their name to complete the sign up process. A new customer record is created with `addCustomer`, and they are then taken into the app.
 - Selecting **Continue as Guest** navigates to the app in guest mode like before.

**Result**
- No longer need the **Registered in person? Set a password** screens from #163 
- No longer need Forgot Password flows (eliminates `PasswordResetScreen`)
- No longer need separate Log In and Sign Up flows (eliminates `LogInScreen`, `SignUpScreen`, and `WelcomeScreen`).

### Additional Updates
- Migrated deprecated [react-native-materialui-textfield](https://github.com/petrogad/react-native-materialui-textfield) to [TextInput from react-native-paper](https://callstack.github.io/react-native-paper/text-input.html) since the original package was no longer supported (issues with `Animated.Text.propTypes`) with the latest React Native upgrades (#170 ).
- Resolved #175 and #122 with the elimination of passwords.

- Converted `RewardsFooter` and `SettingsScreen` to functional components to use the [`useFocusEffect` hook with React Navigation](https://reactnavigation.org/docs/use-focus-effect/). 
    - Original issue: logging out and logging into a different account (or as guest) would not trigger a re-render. In #148, we used a temporary workaround with `Updates.reload()` to force the entire app to reload, but the full reload was very slow, and we weren't able to specify which screen to navigate after the reload.
    - Change: converting these to functional components allows `useFocusEffect` to update the screen when it is in focus.
    - Result: no longer need to use `Updates.reload()` to force refresh. Instead, we use [`CommonActions.reset` from React Navigation](https://reactnavigation.org/docs/navigation-actions/#reset) which resets the navigation state.

## How to review
To test the auth flow:
- Sign up for a new account
- Log in to an existing account that was created on the app
- Log in to an existing account that was created by signing up in-store (through the tablet app)

To test the settings screens (change name and phone number):
- Attempt changing the phone number to a number connected to an existing account
- Attempt changing the phone number to a valid number
- Verify that the user's name updates across the app when changed:
  - Updates in the hamburger menu
  - Updates in the settings screen


## Relevant Links

### Online sources
The redesigned password-less flow was inspired by the auth flow of Uber and Lyft.

[React: How do I bind a function to a component instance?](https://reactjs.org/docs/faq-functions.html#how-do-i-bind-a-function-to-a-component-instance)
https://callstack.github.io/react-native-paper/text-input.html

## Next steps
- Update text field styling to match the rest of the app.
- Update documentation

### Screenshots
https://www.loom.com/share/c5eb405b355c4204bda071ba0de4a936
